### ISSUE TITLE
Cannot build APK - IndexError: List index out of range

### ISSUE THREAD

**Issue by myasul** [kivy/python-for-android#1774]
### Versions

* Python: 3.6.7
* OS: Ubuntu 18.04.2 LTS
* Buildozer:

### Description

I tried building my kivy app via buildozer. There are a lot of errors but stackoverflow and previous helped me up to this point. I'm currently stuck. I already tried rebuilding (deleting .buildozer folder then running buildozer -v android debug but I still encounter the same issue. Please advise.

### buildozer.spec
Command:
```sh
buidozer -v android debug
```

Spec file:
```
[app]

# (str) Title of your application
title = EKT Dictionary

# (str) Package name
package.name = ekt_dictionary

# (str) Package domain (needed for android/ios packaging)
package.domain = org.test

# (str) Source code where the main.py live
source.dir = .

# (list) Source files to include (let empty to include all the files)
source.include_exts =

# (list) List of inclusions using pattern matching
#source.include_patterns = assets/*,images/*.png

# (list) Source files to exclude (let empty to not exclude anything)
#source.exclude_exts = spec

# (list) List of directory to exclude (let empty to not exclude anything)
#source.exclude_dirs = tests, bin

# (list) List of exclusions using pattern matching
#source.exclude_patterns = license,images/*/*.jpg

# (str) Application versioning (method 1)
version = 0.1

# (str) Application versioning (method 2)
# version.regex = __version__ = ['"](.*)['"]
# version.filename = %(source.dir)s/main.py

# (list) Application requirements
# comma separated e.g. requirements = sqlite3,kivy
requirements = libffi,python3,kivy,sqlalchemy

# (str) Custom source folders for requirements
# Sets custom source for any requirements with recipes
# requirements.source.kivy = ../../kivy

# (list) Garden requirements
#garden_requirements =

# (str) Presplash of the application
#presplash.filename = %(source.dir)s/data/presplash.png

# (str) Icon of the application
#icon.filename = %(source.dir)s/data/icon.png

# (str) Supported orientation (one of landscape, sensorLandscape, portrait or all)
orientation = portrait

# (list) List of service to declare
#services = NAME:ENTRYPOINT_TO_PY,NAME2:ENTRYPOINT2_TO_PY

#
# OSX Specific
#

#
# author = © Copyright Info

# change the major version of python used by the app
osx.python_version = 3

# Kivy version to use
osx.kivy_version = 1.10.1

#
# Android specific
#

# (bool) Indicate if the application should be fullscreen or not
fullscreen = 0

# (string) Presplash background color (for new android toolchain)
# Supported formats are: #RRGGBB #AARRGGBB or one of the following names:
# red, blue, green, black, white, gray, cyan, magenta, yellow, lightgray,
# darkgray, grey, lightgrey, darkgrey, aqua, fuchsia, lime, maroon, navy,
# olive, purple, silver, teal.
#android.presplash_color = #FFFFFF

# (list) Permissions
#android.permissions = INTERNET

# (int) Target Android API, should be as high as possible.
#android.api = 27

# (int) Minimum API your APK will support.
#android.minapi = 21

# (int) Android SDK version to use
#android.sdk = 20

# (str) Android NDK version to use
#android.ndk = 17c

# (int) Android NDK API to use. This is the minimum API your app will support, it should usually match android.minapi.
#android.ndk_api = 21

# (bool) Use --private data storage (True) or --dir public storage (False)
#android.private_storage = True

# (str) Android NDK directory (if empty, it will be automatically downloaded.)
#android.ndk_path =

# (str) Android SDK directory (if empty, it will be automatically downloaded.)
#android.sdk_path =

# (str) ANT directory (if empty, it will be automatically downloaded.)
#android.ant_path =

# (bool) If True, then skip trying to update the Android sdk
# This can be useful to avoid excess Internet downloads or save time
# when an update is due and you just want to test/build your package
# android.skip_update = False

# (bool) If True, then automatically accept SDK license
# agreements. This is intended for automation only. If set to False,
# the default, you will be shown the license when first running
# buildozer.
# android.accept_sdk_license = False

# (str) Android entry point, default is ok for Kivy-based app
#android.entrypoint = org.renpy.android.PythonActivity

# (list) Pattern to whitelist for the whole project
#android.whitelist =

# (str) Path to a custom whitelist file
#android.whitelist_src =

# (str) Path to a custom blacklist file
#android.blacklist_src =

# (list) List of Java .jar files to add to the libs so that pyjnius can access
# their classes. Don't add jars that you do not need, since extra jars can slow
# down the build process. Allows wildcards matching, for example:
# OUYA-ODK/libs/*.jar
#android.add_jars = foo.jar,bar.jar,path/to/more/*.jar

# (list) List of Java files to add to the android project (can be java or a
# directory containing the files)
#android.add_src =

# (list) Android AAR archives to add (currently works only with sdl2_gradle
# bootstrap)
#android.add_aars =

# (list) Gradle dependencies to add (currently works only with sdl2_gradle
# bootstrap)
#android.gradle_dependencies =

# (list) Java classes to add as activities to the manifest.
#android.add_activites = com.example.ExampleActivity

# (str) python-for-android branch to use, defaults to master
p4a.branch = master

# (str) OUYA Console category. Should be one of GAME or APP
# If you leave this blank, OUYA support will not be enabled
#android.ouya.category = GAME

# (str) Filename of OUYA Console icon. It must be a 732x412 png image.
#android.ouya.icon.filename = %(source.dir)s/data/ouya_icon.png

# (str) XML file to include as an intent filters in <activity> tag
#android.manifest.intent_filters =

# (str) launchMode to set for the main activity
#android.manifest.launch_mode = standard

# (list) Android additional libraries to copy into libs/armeabi
#android.add_libs_armeabi = libs/android/*.so
#android.add_libs_armeabi_v7a = libs/android-v7/*.so
#android.add_libs_x86 = libs/android-x86/*.so
#android.add_libs_mips = libs/android-mips/*.so

# (bool) Indicate whether the screen should stay on
# Don't forget to add the WAKE_LOCK permission if you set this to True
#android.wakelock = False

# (list) Android application meta-data to set (key=value format)
#android.meta_data =

# (list) Android library project to add (will be added in the
# project.properties automatically.)
#android.library_references =

# (str) Android logcat filters to use
#android.logcat_filters = *:S python:D

# (bool) Copy library instead of making a libpymodules.so
#android.copy_libs = 1

# (str) The Android arch to build for, choices: armeabi-v7a, arm64-v8a, x86, x86_64
android.arch = armeabi-v7a

#
# Python for android (p4a) specific
#

# (str) python-for-android git clone directory (if empty, it will be automatically cloned from github)
#p4a.source_dir =

# (str) The directory in which python-for-android should look for your own build recipes (if any)
#p4a.local_recipes =

# (str) Filename to the hook for p4a
#p4a.hook =

# (str) Bootstrap to use for android builds
# p4a.bootstrap = sdl2

# (int) port number to specify an explicit --port= p4a argument (eg for bootstrap flask)
#p4a.port =


#
# iOS specific
#

# (str) Path to a custom kivy-ios folder
#ios.kivy_ios_dir = ../kivy-ios
# Alternately, specify the URL and branch of a git checkout:
ios.kivy_ios_url = https://github.com/kivy/kivy-ios
ios.kivy_ios_branch = master

# Another platform dependency: ios-deploy
# Uncomment to use a custom checkout
#ios.ios_deploy_dir = ../ios_deploy
# Or specify URL and branch
ios.ios_deploy_url = https://github.com/phonegap/ios-deploy
ios.ios_deploy_branch = 1.7.0

# (str) Name of the certificate to use for signing the debug version
# Get a list of available identities: buildozer ios list_identities
#ios.codesign.debug = "iPhone Developer: <lastname> <firstname> (<hexstring>)"

# (str) Name of the certificate to use for signing the release version
#ios.codesign.release = %(ios.codesign.debug)s


[buildozer]

# (int) Log level (0 = error only, 1 = info, 2 = debug (with command output))
log_level = 2

# (int) Display warning if buildozer is run as root (0 = False, 1 = True)
warn_on_root = 1

# (str) Path to build artifact storage, absolute or relative to spec file
# build_dir = ./.buildozer

# (str) Path to build output (i.e. .apk, .ipa) storage
# bin_dir = ./bin

#    -----------------------------------------------------------------------------
#    List as sections
#
#    You can define all the "list" as [section:key].
#    Each line will be considered as a option to the list.
#    Let's take [app] / source.exclude_patterns.
#    Instead of doing:
#
#[app]
#source.exclude_patterns = license,data/audio/*.wav,data/images/original/*
#
#    This can be translated into:
#
#[app:source.exclude_patterns]
#license
#data/audio/*.wav
#data/images/original/*
#


#    -----------------------------------------------------------------------------
#    Profiles
#
#    You can extend section / key with a profile
#    For example, you want to deploy a demo version of your application without
#    HD content. You could first change the title to add "(demo)" in the name
#    and extend the excluded directories to remove the HD content.
#
#[app@demo]
#title = My Application (demo)
#
#[app:source.exclude_patterns@demo]
#images/hd/*
#
#    Then, invoke the command line with the "demo" profile:
#
#buildozer --profile demo android debug

```

### Logs

```
[INFO]:    -> running python setup.py install -O2 --root=/home/myasul/Documents/personal_projects/ekt_dictionary/.buildozer/android/platform/build/bu...(and 50 more)
[INFO]:    <- directory context /home/myasul/Documents/personal_projects/ekt_dictionary/.buildozer/android/platform/python-for-android                                          
[INFO]:    Building sqlalchemy for armeabi-v7a
[INFO]:    sqlalchemy apparently isn't already in site-packages
[INFO]:    Building compiled components in sqlalchemy
[INFO]:    -> directory context /home/myasul/Documents/personal_projects/ekt_dictionary/.buildozer/android/platform/build/build/other_builds/sqlalchemy-python3/armeabi-v7a__ndk_target_21/sqlalchemy
[INFO]:    -> running python setup.py build_ext -v
Traceback (most recent call last):                                                                                                                                              
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/myasul/Documents/personal_projects/ekt_dictionary/.buildozer/android/platform/python-for-android/pythonforandroid/toolchain.py", line 1081, in <module>
    main()
  File "/home/myasul/Documents/personal_projects/ekt_dictionary/.buildozer/android/platform/python-for-android/pythonforandroid/toolchain.py", line 1075, in main
    ToolchainCL()
  File "/home/myasul/Documents/personal_projects/ekt_dictionary/.buildozer/android/platform/python-for-android/pythonforandroid/toolchain.py", line 577, in __init__
    getattr(self, args.subparser_name.replace('-', '_'))(args)
  File "/home/myasul/Documents/personal_projects/ekt_dictionary/.buildozer/android/platform/python-for-android/pythonforandroid/toolchain.py", line 151, in wrapper_func
    build_dist_from_args(ctx, dist, args)
  File "/home/myasul/Documents/personal_projects/ekt_dictionary/.buildozer/android/platform/python-for-android/pythonforandroid/toolchain.py", line 200, in build_dist_from_args
    build_recipes(build_order, python_modules, ctx)
  File "/home/myasul/Documents/personal_projects/ekt_dictionary/.buildozer/android/platform/python-for-android/pythonforandroid/build.py", line 562, in build_recipes
    recipe.build_arch(arch)
  File "/home/myasul/Documents/personal_projects/ekt_dictionary/.buildozer/android/platform/python-for-android/pythonforandroid/recipe.py", line 899, in build_arch
    self.build_compiled_components(arch)
  File "/home/myasul/Documents/personal_projects/ekt_dictionary/.buildozer/android/platform/python-for-android/pythonforandroid/recipe.py", line 912, in build_compiled_components
    build_dir = glob.glob('build/lib.*')[0]
IndexError: list index out of range
```

**Comment by AndreMiras** [comment]
@myasul so this is a confirmed p4a bug, most likely on the [pythonforandroid/recipes/sqlalchemy/__init__.py](https://github.com/kivy/python-for-android/blob/3cc0657/pythonforandroid/recipes/sqlalchemy/__init__.py) recipe. Would you like to give it a try fixing it? I can assist.

**Comment by strubbi77** [comment]
I would be very interested that this bug will be fixed soon. 
Any base direction what I should fix?

**Comment by AndreMiras** [comment]
Great that you're willing to take a look :smile: 
So I imagine the idea is that this line `build_dir = glob.glob('build/lib.*')[0]` failed because no `lib.*` were found after `sqlalchemy` got built.
So I would look in your build folder `~/.local/share/python-for-android/build/other_builds/sqlalchemy-python3/armeabi-v7a__ndk_target_21/sqlalchemy` see what was actually built. If it was built properly, but just the name is not matching then you can try to override the `build_compiled_components()` method to fix that part.

**Comment by strubbi77** [comment]
Strange, I get another error using latest master: 
Last time 1-2 months ago I had this glob.glob error as well. 

[INFO]:    -> directory context /home/pille/.local/share/python-for-android/build/other_builds/sqlalchemy-python3/armeabi-v7a__ndk_target_21/sqlalchemy
[INFO]:    -> running python setup.py install -O2 --root=/home/pille/.local/share/python-for-andr...(and 53 more)
           working: ModuleNotFoundError: No module named 'setuptools'                                                       Exception in thread background thread for pid 30446:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.7/dist-packages/sh.py", line 2170, in background_thread
    handle_exit_code(exit_code)
  File "/usr/local/lib/python3.7/dist-packages/sh.py", line 1929, in fn
    return self.command.handle_command_exit_code(exit_code)
  File "/usr/local/lib/python3.7/dist-packages/sh.py", line 672, in handle_command_exit_code
    raise exc
sh.ErrorReturnCode_1: 

  RAN: /home/pille/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/native-build/python setup.py install -O2 --root=/home/pille/.local/share/python-for-android/build/python-installs/MVO_Bed_App --install-lib=.

  STDOUT:
Traceback (most recent call last):
  File "setup.py", line 10, in <module>
    from setuptools import Distribution as _Distribution
ModuleNotFoundError: No module named 'setuptools'


  STDERR:


Traceback (most recent call last):                                                                                          
  File "/usr/local/bin/p4a", line 11, in <module>
    load_entry_point('python-for-android==0.7.1', 'console_scripts', 'p4a')()
  File "/usr/local/lib/python3.7/dist-packages/python_for_android-0.7.1-py3.7.egg/pythonforandroid/toolchain.py", line 1075, in main
    ToolchainCL()
  File "/usr/local/lib/python3.7/dist-packages/python_for_android-0.7.1-py3.7.egg/pythonforandroid/toolchain.py", line 577, in __init__
    getattr(self, args.subparser_name.replace('-', '_'))(args)
  File "/usr/local/lib/python3.7/dist-packages/python_for_android-0.7.1-py3.7.egg/pythonforandroid/toolchain.py", line 151, in wrapper_func
    build_dist_from_args(ctx, dist, args)
  File "/usr/local/lib/python3.7/dist-packages/python_for_android-0.7.1-py3.7.egg/pythonforandroid/toolchain.py", line 200, in build_dist_from_args
    build_recipes(build_order, python_modules, ctx)
  File "/usr/local/lib/python3.7/dist-packages/python_for_android-0.7.1-py3.7.egg/pythonforandroid/build.py", line 562, in build_recipes
    recipe.build_arch(arch)
  File "/usr/local/lib/python3.7/dist-packages/python_for_android-0.7.1-py3.7.egg/pythonforandroid/recipe.py", line 834, in build_arch
    self.install_python_package(arch)
  File "/usr/local/lib/python3.7/dist-packages/python_for_android-0.7.1-py3.7.egg/pythonforandroid/recipe.py", line 855, in install_python_package
    _env=hpenv, *self.setup_extra_args)
  File "/usr/local/lib/python3.7/dist-packages/python_for_android-0.7.1-py3.7.egg/pythonforandroid/logger.py", line 178, in shprint
    for line in output:
  File "/usr/local/lib/python3.7/dist-packages/sh.py", line 720, in next
    self.wait()
  File "/usr/local/lib/python3.7/dist-packages/sh.py", line 651, in wait
    self.handle_command_exit_code(exit_code)
  File "/usr/local/lib/python3.7/dist-packages/sh.py", line 672, in handle_command_exit_code
    raise exc
sh.ErrorReturnCode_1: 

  RAN: /home/pille/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/native-build/python setup.py install -O2 --root=/home/pille/.local/share/python-for-android/build/python-installs/MVO_Bed_App --install-lib=.

  STDOUT:
Traceback (most recent call last):
  File "setup.py", line 10, in <module>
    from setuptools import Distribution as _Distribution
ModuleNotFoundError: No module named 'setuptools'


  STDERR:

**Comment by strubbi77** [comment]
Ran the same in docker. 

  File "/home/user/pythonforandroid/recipe.py", line 912, in build_compiled_components
    build_dir = glob.glob('build/lib.*')[0]
IndexError: list index out of range

But no lib.* files. 
(venv) user@0924040531b1:~/.local/share/python-for-android/build/other_builds/sqlalchemy-python3/armeabi-v7a__ndk_target_21/sqlalchemy$ find build/

build/
build/temp.linux-x86_64-3.7
build/temp.linux-x86_64-3.7/lib
build/temp.linux-x86_64-3.7/lib/sqlalchemy
build/temp.linux-x86_64-3.7/lib/sqlalchemy/cextension

**Comment by myasul** [comment]
@AndreMiras thank you for the quick response. I'll have a look at it tomorrow.

**Comment by strubbi77** [comment]
Should be fixed with #1794 as well.

**Comment by AndreMiras** [comment]
True that :tada:
